geode-core/src/main/java/org/apache/geode/internal/net/NioSslEngine.java

```diff
@@ -0,0 +1,353 @@
+      throw new IllegalStateException("NioSslEngine has been closed");
+    }
+  }
+
+  void handleBlockingTasks() {
+    Runnable task;
+    while ((task = engine.getDelegatedTask()) != null) {
+      // these tasks could be run in other threads but the SSLEngine will block until they finish
+      task.run();
+    }
+  }
+
+  @Override
+  public synchronized ByteBuffer wrap(ByteBuffer appData) throws IOException {
+    checkClosed();
+
+    myNetData.clear();
+
+    while (appData.hasRemaining()) {
+      // ensure we have lots of capacity since encrypted data might
+      // be larger than the app data
+      int remaining = myNetData.capacity() - myNetData.position();
+
+      if (remaining < (appData.remaining() * 2)) {
+        int newCapacity = expandedCapacity(appData, myNetData);
+        myNetData = expandBuffer(Buffers.BufferType.TRACKED_SENDER, myNetData, newCapacity, stats);
+      }
+
+      SSLEngineResult wrapResult = engine.wrap(appData, myNetData);
+
+      if (wrapResult.getHandshakeStatus() == NEED_TASK) {
+        handleBlockingTasks();
+      }
+
+      if (wrapResult.getStatus() != OK) {
+        throw new SSLException("Error encrypting data: " + wrapResult);
+      }
+    }
+
+    myNetData.flip();
+
+    return myNetData;
+  }
+
+  @Override
+  public synchronized ByteBuffer unwrap(ByteBuffer wrappedBuffer) throws IOException {
+    checkClosed();
+
+    // note that we do not clear peerAppData as it may hold a partial
+    // message. TcpConduit, for instance, uses message chunking to
+    // transmit large payloads and we may have read a partial chunk
+    // during the previous unwrap
+    while (wrappedBuffer.hasRemaining()) {
+      int remaining = peerAppData.capacity() - peerAppData.position();
+      if (remaining < wrappedBuffer.remaining() * 2) {
+        int newCapacity = expandedCapacity(peerAppData, wrappedBuffer);
+        peerAppData = expandBuffer(Buffers.BufferType.UNTRACKED, peerAppData, newCapacity, stats);
+      }
+      SSLEngineResult unwrapResult = engine.unwrap(wrappedBuffer, peerAppData);
+      if (unwrapResult.getStatus() == BUFFER_UNDERFLOW) {
+        // partial data - need to read more. When this happens the SSLEngine will not have
+        // changed the buffer position
+        wrappedBuffer.compact();
+        return peerAppData;
+      }
+
+      if (unwrapResult.getStatus() != OK) {
+        throw new SSLException("Error decrypting data: " + unwrapResult);
+      }
+    }
+    wrappedBuffer.clear();
+    return peerAppData;
+  }
+
+  @Override
+  public ByteBuffer getUnwrappedBuffer(ByteBuffer wrappedBuffer) {
+    return peerAppData;
+  }
+
+  @Override
+  public ByteBuffer ensureUnwrappedCapacity(int amount, ByteBuffer wrappedBuffer,
+      Buffers.BufferType bufferType,
+      DMStats stats) {
+    peerAppData = Buffers.expandBuffer(bufferType, peerAppData, amount, this.stats);
+    return peerAppData;
+  }
+
+
+  @Override
+  public void close(SocketChannel socketChannel) {
```

Comment:
Will `NioSslEngine.close()` block if the socket is hung somehow due to network weirdness? Maybe that's fine and I don't want to block the PR for this, but I wanted to solicit your opinion.

Reply:
Yes - even socket.close() can block when there are network problems.  That's why we usually spin off AsyncCloser threads to close sockets.